core/lib/include/compiler_hints.h

```diff
@@ -164,6 +165,22 @@ extern "C" {
  */
 #define unlikely(x)     __builtin_expect((uintptr_t)(x), 0)
 
+/**
+ * @brief   Behaves like an `assert()`, but tells the compiler that @p cond can
+ *          never be false.
+ *          This allows the compiler to optimize the code accordingly even when
+ *          `NDEBUG` is set / with `DEVELHELP=0`.
+ *
+ *          @p cond being false will result in undefined behavior.
+ *
+ * @param[in] cond  Condition that is guaranteed to be true
+ */
+#ifdef NDEBUG
+#define assume(cond)    ((cond) ? (void)0 : __builtin_unreachable())
```

Comment:
```suggestion
#define assume(cond)    ((cond) ? (void)0 : UNREACHABLE())
```

Not all compilers have that built-in, but that's already taken care of in line 94.

Reply:
that check doesn't work with `TOOLCHAIN=llvm` :confused:

Reply:
Well then maybe our UNREACHABLE macro needs an update?